src/vodozemac/ecies.rs

```diff
@@ -0,0 +1,209 @@
+#[wasm_bindgen(getter_with_clone)]
+pub struct OutboundCreationResult {
+    /// The established ECIES channel.
+    pub channel: EstablishedEcies,
+    /// The initial encrypted message.
+    pub initial_message: String,
+}
+
+/// An unestablished ECIES session.
+#[wasm_bindgen]
+pub struct Ecies {
+    inner: Option<ecies::Ecies>,
+    public_key: Curve25519PublicKey,
+}
+
+#[wasm_bindgen]
+impl Ecies {
+    /// Create a new, random, unestablished ECIES session.
+    ///
+    /// This method will use the `MATRIX_QR_CODE_LOGIN` info.
+    #[wasm_bindgen(constructor)]
+    pub fn new() -> Self {
+        let inner = ecies::Ecies::new();
+        let public_key = inner.public_key().into();
+
+        Self { inner: Some(inner), public_key }
+    }
+
+    /// Get our [`Curve25519PublicKey`].
+    ///
+    /// This public key needs to be sent to the other side to be able to
+    /// establish an ECIES channel.
+    pub fn public_key(&self) -> Curve25519PublicKey {
+        self.public_key.clone()
+    }
+
+    fn used_up_error() -> JsError {
+        JsError::new("The ECIES channel was already established and used up.")
+    }
+
+    /// Create a [`EstablishedEcies`] from an initial message encrypted by the
+    /// other side.
+    pub fn establish_inbound_channel(
+        &mut self,
+        initial_message: &str,
+    ) -> Result<InboundCreationResult, JsError> {
+        let message = ecies::InitialMessage::decode(&initial_message)?;
+        let result = self
+            .inner
+            .take()
+            .ok_or_else(Self::used_up_error)?
+            .establish_inbound_channel(&message)?;
+
+        let message = String::from_utf8_lossy(&result.message).to_string();
+
+        Ok(InboundCreationResult { message, channel: result.ecies.into() })
+    }
+
+    /// Create an [`EstablishedEcies`] session using the other side's Curve25519
+    /// public key and an initial plaintext.
+    ///
+    /// After the channel has been established, we can encrypt messages to send
+    /// to the other side. The other side uses the initial message to
+    /// establishes the same channel on its side.
+    pub fn establish_outbound_channel(
+        &mut self,
+        public_key: &Curve25519PublicKey,
+        initial_message: &str,
+    ) -> Result<OutboundCreationResult, JsError> {
+        let result = self
+            .inner
+            .take()
+            .ok_or_else(Self::used_up_error)?
+            .establish_outbound_channel(public_key.inner, initial_message.as_bytes())?;
+
+        Ok(OutboundCreationResult {
+            initial_message: result.message.encode(),
+            channel: result.ecies.into(),
+        })
+    }
+}
+
+/// An established ECIES session.
+///
+/// This session can be used to encrypt and decrypt messages between the two
+/// sides of the channel.
+#[derive(Clone)]
+#[wasm_bindgen]
+pub struct EstablishedEcies {
+    inner: Arc<Mutex<ecies::EstablishedEcies>>,
```

Comment:
Out of curiosity, are the arc/mutex needed here, since this is wasm and in theory mono-threaded? (unless the consumer is making use of shared array buffers)

Reply:
AFAIR yes, leaving them out did not work. On WASM they will be compiled into a no-op, but the Rust compiler complains about... Err I forgot when/where it complained.